Implement feature/SPF2-18: Warehouse Edit API

davidmaxxxy opened the pull request:
Implemented the backend API for editing warehouse details.

Changes made:

- Created the `updateWarehouseDetails` controller in `warehouseController.js` to handle PUT requests.

- Added data validation using Joi to ensure non-empty values and proper format.

- Integrated the new routes in `warehouseRoutes.js`.

davidmaxxxy committed:
Implement feature/SPF2-18: Warehouse Edit API

Implemented the backend API for editing warehouse details.

Changes made:

- Created the `updateWarehouseDetails` controller in `warehouseController.js` to handle PUT requests.

- Added data validation using Joi to ensure non-empty values and proper format.

- Integrated the new routes in `warehouseRoutes.js`.

bernard493 commented on server.js:
```diff
@@ -1,27 +1,23 @@
-app.use('/inventories/:warehouse_id', inventoriesRoutes);
-
+// Routes
+app.use("/inventories/", inventoriesRoutes);
```
why did you change this line of code from app.use('/inventories/:warehouse_id', inventoriesRoutes);  to  app.use("/inventories/", inventoriesRoutes);

davidmaxxxy replied:
Unintentionally. Probably changed it whilst dealing with warehouses routes.

bernard493 replied:
oh ok then you need to change it back cox to add an inventory you need to be able to know what warehouse that inventory belongs to that why i added the warehouse_id to the route

bernard493 commented on server.js:
```diff
@@ -1,27 +1,23 @@
-
+// Routes
+app.use("/inventories/", inventoriesRoutes);
+app.use("/warehouses/", warehouseRoutes);
```
In your warehouseController, you get the warehouse_id from the req.params.id but in this server js file your route is not dynamically allowing a params in the route, which means req.params.id will be undefined  
and if you check your ticket on Jira you gonna see the expected route 

Fix 
```
server.js
app.use("/warehouses/:id", warehouseRoutes);

warehouseController.js
const warehouseId = parseInt(req.params.id);
```


If you modify your route to app.use("/warehouses/:id", warehouseRoutes);, then the API endpoint will only be accessible when the request includes a valid warehouse ID, such as warehouses/554421.

bernard493 commented on controllers/warehouseController.js:
```diff
@@ -0,0 +1,58 @@
+    if (!warehouseId || isNaN(warehouseId)) {
+      return res.status(400).json({ message: "Invalid warehouse ID." });
+    }
+
```
If you modify your route to app.use("/warehouses/:id", warehouseRoutes);, then the API endpoint will only be accessible when the request includes a valid warehouse ID, such as warehouses/554421. This eliminates the need to manually check if the ID was passed or if it's invalid in the controller.

bernard493 commented on controllers/warehouseController.js:
```diff
@@ -0,0 +1,58 @@
+    const updatedWarehouse = await knex("warehouses")
+      .where("id", warehouseId)
+      .first();
+
```
You are retrieving the updated warehouse after executing the update query, but there is a chance the update operation could fail, and you are not checking whether it was successful. I recommend adding a check to ensure the update was successful before querying the database for the updated information. If the update fails, you can return an appropriate error response instead of making an additional DB call.

```

const updateCount   = await knex("warehouses").where("id", warehouseId).update(req.body);

// 0 ==  failed
//1
if (updateCount === 0) {
  return res.status(500).json({ message: "Update failed: Warehouse not found or no changes made. Try again later " });
}
```

**NOTE** 
check updateCount  to make sure its what am thinking

davidmaxxxy committed:
Implement feature/SPF2-18: Warehouse Edit API

Fixed the bugs mentioned by @bernard

Implemented the backend API for editing warehouse details.

Changes made:

- Created the `updateWarehouseDetails` controller in `warehouseController.js` to handle PUT requests.

- Added data validation using Joi to ensure non-empty values and proper format.

- Integrated the new routes in `warehouseRoutes.js`.

davidmaxxxy committed:
feature/SPF2-18: Warehouse Edit API - fixed the bugs with the API not working

I had to remove the :id from the warehouse route Ensure that in server.js to ensure that I was setting the general route (/warehouses), not the specific route with :id. Otherwise it was returning and error. The PUT route for /warehouses/:id should remain inside warehouseRoutes.js.
